StepContainerV2: Fix blinking and Flash Of Wrong Viewport

zaguiini opened the pull request:
Supersedes https://github.com/Automattic/wp-calypso/pull/101944.

## Proposed Changes

### Blinking

We were getting blinks in two different stages:

#### Entering `/setup`

The `getFlowFromURL()` call within `pages/index.js` returns an empty string, which causes `shouldUseStepContainerV2` to return `false` as the flow it was comparing against was `''`.

Now, if we know we're entering `setup` and no flow is detected, we're defaulting against `onboarding`, which is our default flow.

#### Entering `/setup/onboarding`

Within the flow renderer we're using a fallback route provided by `react-router` to redirect to the first step. This is not a problem. The problem is that the redirect logic uses the `<Navigate />` component, which is a [`useEffect wrapper` call](https://github.com/remix-run/react-router/blob/a3e4b8ed875611637357647fcf862c2bc61f4e11/packages/react-router/lib/components.tsx#L570-L572).

That `useEffect` call (instead of `useLayoutEffect`) caused one (empty) render before the redirect, making the top bar "blink".

#### Entering `/setup/onboarding/domains`

There was no blink here as we have both a matching flow and a first step.

### Flash Of Wrong Viewport

The Calypso boot process happens on the server, and `StepContainerV2` relies on screen measures that are only possible to be done in the client. Here's what was happening:

1. User on desktop requested the page;
2. Server processed the `StepContainerV2Loading` component and returned the markup (wrong viewport);
3. Markup gets rendered in the client (wrong viewport);
4. Markup gets hydrated (with the correct viewport).

Switching to static CSS media queries instead of the value returned from `useViewportMatch` (which doesn't have an effect on the server) fixed that problem. In https://github.com/Automattic/wp-calypso/pull/101985, we're going to move away from `useViewportMatch` entirely.

## Testing instructions

Enter `/setup`, `/setup/onboarding` and `/setup/onboarding/domains` and verify you don't see the top bar blink or change from the W logo to the wordmark.

## Recordings

### Before

https://github.com/user-attachments/assets/bce580ae-741b-4472-82fc-e6ade6eec4c3

### After

https://github.com/user-attachments/assets/ad3037a7-1057-4be8-b969-ba0c0013cab8

p-jackson committed:
Server call to getFlowFromURL returns correct value

zaguiini committed:
Redirect to first step before committing to render within Stepper

zaguiini committed:
Use onboarding as the default flow when entering /setup directly

zaguiini committed:
Statically render wordmark or logo using a media query instead of viewport match in JS

zaguiini committed:
Extract DEFAULT_FLOW to its own variable

zaguiini committed:
Fix tests

zaguiini commented on client/landing/stepper/declarative-flow/internals/components/redirect-to-step/test/index.tsx:
This is weird to me. It started failing after I switched the component from `useEffect` to `useLayoutEffect`, and now even with `waitFor` it doesn't really pick up the changes. I'm guessing because the layout effect is sync it's not detecting the changes and not queueing in RTL. 

Anyway, this is alright for the moment as I'm more worried in deploying this. cc @p-jackson if you have time for a follow up 🤞🏻

p-jackson replied:
Ok, so it seems a bit daft, but it doesn't look like you can use `navigate` within `useLayoutEffect`

https://github.com/remix-run/react-router/issues/10941#issuecomment-1769329661

It's working for us right now. Maybe it's something that'll bite us down the line though? Because it's like it just happens to work right now in `BrowserRouter` but not in `MemoryRouter` (used by the tests).

Not ideal.

p-jackson replied:
I believe what the React Router dev is trying to say is that ideally you wouldn't need to do synchronous redirects, because if you did, you'd do those redirects on the server. (That's the `loader` they're talking about, pretty sure that's the `loader` function from the Remix library).

It is true. Redirecting on the server probably would be better. But I won't do that, because I think most Calypso devs would find that unexpected and hard to debug if they ever needed to figure it out.

The solution I believe is to just render the appropriate components while we wait for the redirect to happen.

[`b4f1748` (#101967)](https://github.com/Automattic/wp-calypso/pull/101967/commits/b4f1748b31d27927f8dbc326871777f39174befa#diff-1a5cb229a76dfccdc8b471efedc2ab09f2385c9ba00ae3211ba65452ba1c7680R244-R249)

zaguiini commented on packages/onboarding/src/step-container-v2/components/TopBar/style.scss:
In https://github.com/Automattic/wp-calypso/pull/101985, we're going to refactor the remaining components. The changes to this file have been ported from there.

paulopmt1 commented on client/landing/stepper/utils/get-flow-from-url.ts:
```diff
@@ -1,9 +1,12 @@
+export const DEFAULT_FLOW = ONBOARDING_FLOW;
+
+export const getFlowFromURL = ( pathname?: string, search?: string ) => {
+	pathname ??= typeof window !== 'undefined' ? window.location.pathname ?? '' : '';
```
Took me a minute to break this down 😆 

But I see: we want to only define pathname if it's not defined already, and we'll set it as `window.location.pathname` otherwise (unless that's also not defined and if we do have the window object).

p-jackson replied:
Yeah, it's a ridiculous number of `?`s 🙈 I feel like some explicit parenthesis would make this easier to read. But Prettier insists on removing them.

paulopmt1 commented on client/landing/stepper/utils/test/get-flow-from-url.ts:
```diff
@@ -0,0 +1,19 @@
+import { getFlowFromURL } from '../get-flow-from-url';
```
Thanks for making the improvements by adding unit tests!

zaguiini replied:
Courtesy of Mr. Jackson!

p-jackson replied:
I just couldn't figure out why `getFlowFromURL` seemed to be not working properly! Great reminder that unit tests are also a great way to quickly iterate on code so that you don't have to refresh the browser and redo the flow over and over.

paulopmt1 reviewed (approved):
Thanks for fixing this @zaguiini!

Testing the three endpoints generated the same result I'm sharing in this videos:

**Trunk branch - blinking**

https://github.com/user-attachments/assets/4a3781bc-b149-4874-84a0-5f71906c5bd7

**This PR changes - no more initial blinking**


https://github.com/user-attachments/assets/e8623805-6c73-45e7-8eba-813d7d3c1358


We still have two extra blinks in the checkout process when going to the /goals screen. Or one blinking when going to the /goals screen without the checkout. But that can be addressed as a follow-up.

p-jackson commented on client/landing/stepper/declarative-flow/internals/components/redirect-to-step/index.tsx:
```diff
@@ -11,12 +11,20 @@ interface Props {
+	const destination = `${ to }${ window.location.search }`;
+
+	// This needs to be a layout effect, otherwise an empty screen will be shown before the redirect.
+	useLayoutEffect( () => {
+		navigate( destination, { replace: true } );
+	}, [ destination, navigate ] );
```
I wonder if this can be done even more directly be calling `navigate()` directly within the render function ...

p-jackson replied:
Nope that doesn't work. React Router emits a warning about having to be called within an effect. I think internally `useNavigate` is using some sort of ref or something that means it wants to be sure it's been initialised before you call `navigate`

p-jackson committed:
Do not call navigate within useLayoutEffect

This doesn't appear to be supported by React Router. Better to provide a
proper fallback component that we can render until the redirect happens.

p-jackson reviewed (approved):
😌 I think we're good to go. I think I still saw something glitchy when I tested with a logged out user, but I'll try it again and follow up in a different PR. This is already a huge improvement.

autumnfjeld commented:
Tested on calypso live:

I did not see any blinking/flashing

https://www.awesomescreenshot.com/video/38132430?key=5c43a9cfb62524e16ac4c98402f6c6d2
